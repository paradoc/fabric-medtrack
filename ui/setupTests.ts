import '@testing-library/jest-dom/extend-expect'


import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import duration from 'dayjs/plugin/duration'
import relativeTime from 'dayjs/plugin/relativeTime'

dayjs.extend(isSameOrAfter)
dayjs.extend(duration)
dayjs.extend(relativeTime)


export function setupFetchStub(data: any) {
  return function fetchStub(_url: string) {
    return new Promise((resolve) => {
      resolve({
        json: () => Promise.resolve(data),
      })
    })
  }
}