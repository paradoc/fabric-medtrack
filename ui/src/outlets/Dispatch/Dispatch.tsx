import dayjs from 'dayjs'
import React, { useState } from 'react'
import { useForm } from 'react-hook-form'
import { Navigate } from 'react-router'

import { Medication } from '../../views/Collector'

import styles from './Dispatch.module.css'

type SerializedData = {
  medications: Medication[]
  dispatch_date: string
}

type DispatchFormData = {
  generic: string
  brand: string
  frequency: number
  count: number
}

export default function Dispatch() {
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<DispatchFormData>()
  const [dispatchId, setDispatchId] = useState<string>()

  const onSubmit = handleSubmit(
    async ({ brand, count, frequency, generic }) => {
      const preSerialized: SerializedData = {
        medications: [
          {
            brand_name: brand,
            generic_name: generic,
            end_after_n: new Number(count) as number,
            frequency: `PT${frequency}H`,
          },
        ],
        dispatch_date: dayjs().format('YYYY-MM-DD HH:mm:ss'),
      }
      const response = await fetch('/api/dispatch', {
        method: 'POST',
        body: JSON.stringify(preSerialized),
        headers: { 'content-type': 'application/json' },
      })
      const data = await response.json()
      const { dispatch_id } = data
      reset()
      setDispatchId(dispatch_id)
    }
  )

  return dispatchId ? (
    <Navigate to={`../inspect/${dispatchId}`} relative="path" />
  ) : (
    <div className={styles.container}>
      <header>Dispatch</header>
      <form className={styles.form} onSubmit={onSubmit}>
        <div className={styles.inputField}>
          <label htmlFor="generic">Generic</label>
          <input
            id="generic"
            type="text"
            {...register('generic', { required: true })}
          />
          <span>{errors.generic && '* required'}</span>
        </div>

        <div className={styles.inputField}>
          <label htmlFor="brand">Brand</label>
          <input id="brand" {...register('brand', { required: true })} />
          <span>{errors.brand && '* required'}</span>
        </div>

        <div className={styles.inputField}>
          <label htmlFor="frequency">Frequency</label>
          <input
            type="number"
            id="frequency"
            {...register('frequency', { required: true, min: 1, max: 72 })}
            placeholder="in hours"
          />
          <span>{errors.frequency && '* required'}</span>
        </div>

        <div className={styles.inputField}>
          <label htmlFor="count">Count</label>
          <input
            type="number"
            id="count"
            {...register('count', { required: true, min: 1 })}
          />
          <span>{errors.count && '* required'}</span>
        </div>

        <input type="submit" className={styles.submit} value="Dispatch" />
      </form>
    </div>
  )
}
