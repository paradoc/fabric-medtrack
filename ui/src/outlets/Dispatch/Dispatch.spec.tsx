import React from 'react'
import { act } from 'react-dom/test-utils'
import { fireEvent, render } from '@testing-library/react'
import { setupFetchStub } from '../../../setupTests'
import Dispatch from './Dispatch'

describe('Dispatch', () => {
  beforeEach(() => {
    const fakeData = {
      dispatch_id: 'test-dispatch-id',
    }
    global.fetch = jest.fn().mockImplementation(setupFetchStub(fakeData))
  })

  afterEach(() => {
    // global.fetch.mockClear()
    // delete global.fetch
  })

  test('should match snapshot', () => {
    const { container } = render(<Dispatch />)
    act(() => {
      expect(container).toMatchSnapshot()
    })
  })

  test('should check click behavior', async () => {
    const { container, getByTestId } = render(<Dispatch />)
    const submit = getByTestId('submit')
    const brand = getByTestId('brand')
    const frequency = getByTestId('frequency')
    const generic = getByTestId('generic')
    const count = getByTestId('count')
    act(() => {
      fireEvent.change(brand, { target: { value: 'test-brand' } })
      fireEvent.change(generic, { target: { value: 'test-generic' } })
      fireEvent.change(frequency, { target: { value: 2 } })
      fireEvent.change(count, { target: { value: 1 } })
      fireEvent.click(submit)
    })
    expect(container).toMatchSnapshot()
  })
})
