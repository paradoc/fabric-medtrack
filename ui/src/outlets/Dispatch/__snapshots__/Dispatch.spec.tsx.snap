// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispatch should check click behavior 1`] = `
<div>
  <div
    class="container"
  >
    <header>
      Dispatch
    </header>
    <form
      class="form"
    >
      <div
        class="inputField"
      >
        <label
          for="generic"
        >
          Generic
        </label>
        <input
          data-testid="generic"
          id="generic"
          name="generic"
          type="text"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="brand"
        >
          Brand
        </label>
        <input
          data-testid="brand"
          id="brand"
          name="brand"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="frequency"
        >
          Frequency
        </label>
        <input
          data-testid="frequency"
          id="frequency"
          name="frequency"
          placeholder="in hours"
          type="number"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="count"
        >
          Count
        </label>
        <input
          data-testid="count"
          id="count"
          name="count"
          type="number"
        />
        <span />
      </div>
      <input
        class="submit"
        data-testid="submit"
        type="submit"
        value="Dispatch"
      />
    </form>
  </div>
</div>
`;

exports[`Dispatch should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <header>
      Dispatch
    </header>
    <form
      class="form"
    >
      <div
        class="inputField"
      >
        <label
          for="generic"
        >
          Generic
        </label>
        <input
          data-testid="generic"
          id="generic"
          name="generic"
          type="text"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="brand"
        >
          Brand
        </label>
        <input
          data-testid="brand"
          id="brand"
          name="brand"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="frequency"
        >
          Frequency
        </label>
        <input
          data-testid="frequency"
          id="frequency"
          name="frequency"
          placeholder="in hours"
          type="number"
        />
        <span />
      </div>
      <div
        class="inputField"
      >
        <label
          for="count"
        >
          Count
        </label>
        <input
          data-testid="count"
          id="count"
          name="count"
          type="number"
        />
        <span />
      </div>
      <input
        class="submit"
        data-testid="submit"
        type="submit"
        value="Dispatch"
      />
    </form>
  </div>
</div>
`;
